### PR TITLE
Migrate "Layout customization" part of Them & Logo controller

### PR DESCRIPTION
**Pull request by sarjon**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR migrates part where user can customize Front Office theme's page layouts.
| Type?         | new feature
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10578
| How to test?  | You can access new page via url `admin-dev/index.php/improve/design/theme/page-layout-customization`

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/10816)

### CODE REVIEW

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php
Comment (sarjon) [inline src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php]: hopefully i can replace legacy call by service when this PR gets merged => https://github.com/PrestaShop/PrestaShop/pull/10770 since i dont want to add another method without interface to `CacheClearer`. :/

File: src/PrestaShopBundle/Resources/config/services/adapter/admin.yml
Comment (sarjon) [inline src/PrestaShopBundle/Resources/config/services/adapter/admin.yml:17]: class `PrestaShop\PrestaShop\Core\Addon\Theme\ThemeManagerBuilder` even though it's in `Core` it heavily depends on legacy `Db` so i had to do this :cry: :cry:  :cry:

File: src/PrestaShopBundle/Resources/views/Admin/Improve/Design/Theme/customize_page_layouts.twig
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/Improve/Design/Theme/customize_page_layouts.twig]: How about _Admin.Actions_?

File: src/PrestaShopBundle/Resources/views/Admin/Improve/Design/Theme/customize_page_layouts.twig
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/Improve/Design/Theme/customize_page_layouts.twig]: Should be localized in _Admin.Global_.

File: src/PrestaShopBundle/Resources/views/Admin/Improve/Design/Theme/customize_page_layouts.twig
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/Improve/Design/Theme/customize_page_layouts.twig]: Should be localized in _Admin.Global_.

File: src/PrestaShopBundle/Resources/views/Admin/Improve/Design/Theme/customize_page_layouts.twig
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/Improve/Design/Theme/customize_page_layouts.twig]: Should be localized in _Admin.Global_.

File: src/Adapter/Meta/QueryHandler/GetPagesForLayoutCustomizationHandler.php
Comment (matks) [inline src/Adapter/Meta/QueryHandler/GetPagesForLayoutCustomizationHandler.php:60]: You are performing an `array_map` here.
I dont know though which one is faster.

File: src/PrestaShopBundle/Resources/config/services/core/addon.yml
Comment (matks) [inline src/PrestaShopBundle/Resources/config/services/core/addon.yml:46]: What if the shop theme gets changed during the http request handling by php ?
Is it gonna do the following scenario:
1) SF container loads, gets the shop theme A from context
2) controller gets called, handles http requests and replaces shop theme A, now it's shop theme B
3) another processing is performed which requires to use se Customizer but customization is performed using theme A and not the latest, theme B

File: src/Core/Domain/Meta/DataTransferObject/LayoutCustomizationPage.php
Comment (matks) [inline src/Core/Domain/Meta/DataTransferObject/LayoutCustomizationPage.php:27]: Is it the 1st time you use DataTransferObjects ? Do you want to start using theme more ?

File: src/Core/Domain/Meta/Query/GetPagesForLayoutCustomization.php
Comment (matks) [inline src/Core/Domain/Meta/Query/GetPagesForLayoutCustomization.php:33]: I guess this will be completed in another PR 😄 ?

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php
Comment (matks) [inline src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php]: Why only `delete` ? Is that from the legacy ?

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php
Comment (matks) [reply src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php]: 👍

File: src/PrestaShopBundle/Resources/config/services/bundle/form/form_factory.yml
Comment (matks) [inline src/PrestaShopBundle/Resources/config/services/bundle/form/form_factory.yml:10]: Same here, I'm concerned about these types of calls as this is dynamic and can change

File: src/PrestaShopBundle/Resources/config/services/adapter/admin.yml
Comment (matks) [reply src/PrestaShopBundle/Resources/config/services/adapter/admin.yml:17]: 😭 
Can we add a `@todo` here ? so that we fix this later ?

I know I bother you a lot with `@todo` statements
But it's really helpful when you want to look for technical debt, 1 way to find some is to look for them.

File: src/PrestaShopBundle/Resources/config/services/adapter/meta.yml
Comment (matks) [inline src/PrestaShopBundle/Resources/config/services/adapter/meta.yml:14]: Same issue here although less dangerous I believe.

File: src/PrestaShopBundle/Resources/config/services/core/addon.yml
Comment (matks) [inline src/PrestaShopBundle/Resources/config/services/core/addon.yml:41]: 👍

**Review by matks — CHANGES_REQUESTED** [review]

File: src/Adapter/Meta/QueryHandler/GetPagesForLayoutCustomizationHandler.php
Comment (sarjon) [reply src/Adapter/Meta/QueryHandler/GetPagesForLayoutCustomizationHandler.php:60]: afaik `foreach` is always faster.

File: src/Core/Domain/Meta/DataTransferObject/LayoutCustomizationPage.php
Comment (sarjon) [reply src/Core/Domain/Meta/DataTransferObject/LayoutCustomizationPage.php:27]: what do you mean? :confused:

File: src/Core/Domain/Meta/Query/GetPagesForLayoutCustomization.php
Comment (sarjon) [reply src/Core/Domain/Meta/Query/GetPagesForLayoutCustomization.php:33]: nope, actually its completed. :trollface: (for real, its completed :smile:)

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php
Comment (sarjon) [reply src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php]: i think we cannot merge this PR, until #10770 is merged?

File: src/PrestaShopBundle/Resources/config/services/adapter/admin.yml
Comment (sarjon) [reply src/PrestaShopBundle/Resources/config/services/adapter/admin.yml:17]: i think that `ThemeManagerBuilder` would be a better place for todo as it has to be refactored not to depend on legacy classes, wdyt?

File: src/PrestaShopBundle/Resources/config/services/core/addon.yml
Comment (sarjon) [reply src/PrestaShopBundle/Resources/config/services/core/addon.yml:46]: hmm, i dont really know if thats an issue or maybe dont understand whole process.

if you were to inject `Context` you would do the same `$context->shop->theme` call and get the same result? But doing call directly in service declaration helps isolate `ThemePageLayoutsCustomizer` from legacy, am i missing something? :thinking:

File: src/PrestaShopBundle/Resources/config/services/adapter/admin.yml
Comment (sarjon) [reply src/PrestaShopBundle/Resources/config/services/adapter/admin.yml:17]: actually i got this idea: what if we make `Adapter` and other legacy services deprecated when declaring? For example:

```yaml
    prestashop.adapter.legacy_db:
        class: 'Db'
        factory: ['Db', 'getInstance']
        deprecated: ~ # we could even add a deprecation message here
```
Doing that, would be another way of telling developers not to depend on them and in addition to that, all deprecation notices would be visible on profiler, wdyt?

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php
Comment (sarjon) [reply src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php]: should be `edit`. :+1:

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php
Comment (matks) [reply src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php]: Yes, we wait for @jolelievre who's at PS Day Warsaw? I'll urge him Monday ;)

File: src/PrestaShopBundle/Resources/config/services/adapter/admin.yml
Comment (matks) [reply src/PrestaShopBundle/Resources/config/services/adapter/admin.yml:17]: The idea is interesting but it is not really deprecated as it's still in use 🤔

File: src/PrestaShopBundle/Resources/config/services/core/addon.yml
Comment (matks) [reply src/PrestaShopBundle/Resources/config/services/core/addon.yml:46]: If `Context` is built correctly (gotta check that 😅 ) then its property `shop.theme` is updated when Shop Theme is modified. So if we call `$context->shop->theme` *after* it has been updated we get the right value.
To do that we must call `$context->shop->theme` at the last moment, just before we use it.

File: src/Core/Domain/Meta/DataTransferObject/LayoutCustomizationPage.php
Comment (matks) [reply src/Core/Domain/Meta/DataTransferObject/LayoutCustomizationPage.php:27]: DataTransferObject is a concept that can be enforced in a php project, the same way Form Factories or Data Handlers are now enforced for SF controllers in migrated PrestaShop.

So I wanted to know if you wanted to introduce the concept in PrestaShop using this PR, and start using them everywhere.

File: src/Core/Domain/Meta/Query/GetPagesForLayoutCustomization.php
Comment (matks) [reply src/Core/Domain/Meta/Query/GetPagesForLayoutCustomization.php:33]: o_O

File: src/Core/Domain/Meta/DataTransferObject/LayoutCustomizationPage.php
Comment (sarjon) [reply src/Core/Domain/Meta/DataTransferObject/LayoutCustomizationPage.php:27]: i dont really think its a "new" concept in PS, there are `Dto`s already, except they are not in explicit namespace. :)

File: src/Core/Domain/Meta/Query/GetPagesForLayoutCustomization.php
Comment (sarjon) [reply src/Core/Domain/Meta/Query/GetPagesForLayoutCustomization.php:33]: here's another example of similar Query https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Domain/SqlManagement/Query/GetDatabaseTablesList.php

you dont always need attributes to have a valid query. :)

File: src/PrestaShopBundle/Resources/config/services/adapter/admin.yml
Comment (sarjon) [reply src/PrestaShopBundle/Resources/config/services/adapter/admin.yml:17]: at least we get notices in profiler when using it, wdyt?

**Comment by sarjon** [comment]
@matks anymore changes needed? :)

File: src/PrestaShopBundle/Resources/config/services/adapter/admin.yml
Comment (matks) [reply src/PrestaShopBundle/Resources/config/services/adapter/admin.yml:17]: Good point 👍

File: src/PrestaShopBundle/Resources/config/services/core/addon.yml
Comment (matks) [reply src/PrestaShopBundle/Resources/config/services/core/addon.yml:46]: Let's go for this and change it later if required 👍

**Review by matks — APPROVED** [review]

**Comment by matks** [comment]
@sarjon Let's go for QA !

**Comment by marionf** [comment]
When I change something and save I stay on the same page with the green message 
Before I was back on Theme & logo page
What do you think @TristanLDD ?

**Comment by sarjon** [comment]
> When I change something and save I stay on the same page with the green message

you are right, but since only one page is migrated i didnt want to redirect to legacy. When we migrate all Theme page, we can keep old behavior.

p.s. i think green message should be kept either way, it was not shown in legacy.

**Comment by TristanLDD** [comment]
@marionf @sarjon  This behavior is ok for me

**Comment by matks** [comment]
Thanks @sarjon

**Pull request merged**
